http-common.c

```diff
@@ -113,36 +113,157 @@ set_method(CURL *ehandle, enum http_method method, struct sized_buffer *body)
       break;
   case HTTP_POST:
       curl_easy_setopt(ehandle, CURLOPT_POST, 1L);
-      //set ptr to payload that will be sent via POST/PUT
-      curl_easy_setopt(ehandle, CURLOPT_POSTFIELDS, body->start);
-      curl_easy_setopt(ehandle, CURLOPT_POSTFIELDSIZE, body->size);
       break;
   case HTTP_PATCH:
       curl_easy_setopt(ehandle, CURLOPT_CUSTOMREQUEST, "PATCH");
-      curl_easy_setopt(ehandle, CURLOPT_POSTFIELDS, body->start);
-      curl_easy_setopt(ehandle, CURLOPT_POSTFIELDSIZE, body->size);
       break;
   case HTTP_PUT:
       curl_easy_setopt(ehandle, CURLOPT_CUSTOMREQUEST, "PUT");
-      curl_easy_setopt(ehandle, CURLOPT_POSTFIELDS, body->start);
-      curl_easy_setopt(ehandle, CURLOPT_POSTFIELDSIZE, body->size);
       break;
   default:
       ERR("Unknown http method (code: %d)", method);
   }
+  
+  if (body && body->start) {
+    //set ptr to payload that will be sent via POST/PUT
+    curl_easy_setopt(ehandle, CURLOPT_POSTFIELDS, body->start);
+    curl_easy_setopt(ehandle, CURLOPT_POSTFIELDSIZE, body->size);
+  }
 }
 
 void
-set_url(CURL *ehandle, char *base_api_url, char endpoint[])
+set_url(CURL *ehandle, char base_api_url[], char endpoint[], va_list *args)
 {
+  //create the url route
+  char url_route[MAX_URL_LEN];
+  int ret = vsnprintf(url_route, sizeof(url_route), endpoint, *args);
+  ASSERT_S(ret < (int)sizeof(url_route), "oob write of url_route");
+
   char base_url[MAX_URL_LEN];
-  int ret = snprintf(base_url, sizeof(base_url), "%s%s", base_api_url, endpoint);
+  ret = snprintf(base_url, sizeof(base_url), "%s%s", base_api_url, url_route);
   ASSERT_S(ret < (int)sizeof(base_url), "Out of bounds write attempt");
 
   CURLcode ecode = curl_easy_setopt(ehandle, CURLOPT_URL, base_url);
   ASSERT_S(CURLE_OK == ecode, curl_easy_strerror(ecode));
 }
 
+static void
+default_cb(void *data)
+{ 
+  return; 
+  (void)data; 
+}
+
+static perform_action
+default_success_cb(
+  void *p_data,
+  enum http_code code,
+  struct sized_buffer *body,
+  struct api_header_s *pairs)
+{ 
+  return ACTION_SUCCESS;
+  (void)p_data; 
+  (void)code; 
+  (void)body;
+  (void)pairs;
+}
+
+static perform_action
+default_retry_cb(
+  void *p_data,
+  enum http_code code,
+  struct sized_buffer *body,
+  struct api_header_s *pairs)
+{
+  return ACTION_RETRY;
+  (void)p_data; 
+  (void)code;
+  (void)body;
+  (void)pairs;
+}
+
+static perform_action 
+default_abort_cb(
+  void *p_data,
+  enum http_code code,
+  struct sized_buffer *body,
+  struct api_header_s *pairs)
+{ 
+  return ACTION_ABORT; 
+  (void)p_data; 
+  (void)code; 
+  (void)body;
+  (void)pairs;
+}
+
+void
+perform_request(
+  struct sized_buffer *body,
```

Comment:
It's better to rename this as `perform_request2` to avoid forcing the existing code to migrate to use this immediately after the merge.

Reply:
Ok, I should rename set_url to set_url2 too, since I merged the url building procedure into it